### PR TITLE
Implement type-ahead in auro-select

### PR DESCRIPTION
**Pull request by jordanjones243**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Review Checklist:

- [ ] My update follows the CONTRIBUTING guidelines of this project
- [ ] I have performed a self-review of my own update

<details>
<summary>
RC Checklist
</summary>

### Browsers

[Browsers Support Guide](https://auro.alaskaair.com/support/browsersSupport)

[Dev demo link](https://alaskaairlines.github.io/auro-formkit/)

Android

- [ ] Chrome
- [ ] Firefox

 iOS

- [ ] Chrome
- [ ] Firefox
- [ ] Safari

Desktop

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Scenarios

- [ ] Validated linked issues with issue reporting team
- [ ] Test coverage report review

[Framework playground ](https://github.com/AlaskaAirlines/auro-framework-playground)

- [ ] Next React
- [ ] SvelteKit

</details><br>
**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

## Summary by Sourcery

Add WAI‑ARIA–aligned type-ahead navigation to auro-select while tightening auro-menu’s public API and event contracts, and documenting the new keyboard behavior.

New Features:
- Introduce buffered type-ahead navigation for auro-select that matches options by displayed text, supports cycling on repeated keys, and exposes a configurable typeaheadTimeoutMs property.
- Add type-ahead usage documentation and an accompanying example to the auro-select customization and keyboard behavior docs.

Enhancements:
- Propagate layout-driven size, shape, and noCheckmark settings from auro-select and auro-combobox to their internal auro-menu instances.
- Align auro-menu’s public methods (e.g., selectByValue, reset, clearSelection, handleCustomEvent) and events to a simpler, documented contract, including refined event detail payloads and loading state notifications.
- Improve robustness of menu and menuoption interaction by handling dynamic option lifecycles, ensuring safe navigation/selection, and expanding public API coverage tests.
- Adjust select keyboard strategy so Space and Escape integrate correctly with the new type-ahead buffer and bib toggling behavior.

Documentation:
- Document auro-select type-ahead behavior, configuration, and keyboard interaction, and add a dedicated type-ahead example to the customization docs and API reference.

Tests:
- Add comprehensive unit tests for auro-select type-ahead behavior, buffer timeout handling, space key semantics, and focus/blur edge cases.
- Extend auro-menu tests to cover public API methods, dynamic option changes, event emissions, and noCheckmark propagation while updating interaction suites for the revised single-select behavior.
- Quarantine legacy 5.9 multiSelect- and MenuService-specific tests and fixtures into a dedicated directory for future multi-select work, with README guidance for reactivation.

### CODE REVIEW

File: components/select/docs/partials/keyboard-behavior/keyEvents.md [hunk 1]
```diff
@@ -196,5 +196,14 @@
         </div>
       </td>
     </tr>
+    <tr>
+      <td>Printable character</td>
+      <td>-</td>
+      <td>Collapsed or Expanded</td>
+      <td>Trigger element</td>
+      <td>
+        Activates type-ahead navigation. Opens the bib if collapsed, then advances the <code>focused</code> option to the first enabled option whose displayed text starts with the buffered keystrokes. Repeated keystrokes within <code>typeaheadTimeoutMs</code> extend the buffer; pressing the same character repeatedly cycles through matching options. See the <a href="customize.html#typeAhead">Type-Ahead</a> section on the Customize page.
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/docs/partials/keyboard-behavior/keyEvents.md]: Addressed.

File: components/select/docs/partials/keyboard-behavior/keyEvents.md [hunk 2]
```diff
@@ -196,5 +196,14 @@
         </div>
       </td>
     </tr>
+    <tr>
+      <td>Printable character</td>
+      <td>-</td>
+      <td>Collapsed or Expanded</td>
+      <td>Trigger element</td>
+      <td>
+        Extends the type-ahead buffer. When the buffer matches an enabled option's displayed text, the <code>focused</code> option advances to that option and the bib opens if it was collapsed. A keystroke that does not match any option leaves the bib state unchanged. Repeated keystrokes within <code>typeaheadTimeoutMs</code> extend the buffer; pressing the same character repeatedly cycles through matching options. See the <a href="customize.html#typeAhead">Type-Ahead</a> section on the Customize page.
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/docs/partials/keyboard-behavior/keyEvents.md:214]: Addressed.

File: components/select/src/auro-select.js [hunk 3]
```diff
@@ -880,41 +902,86 @@ export class AuroSelect extends AuroElement {
...
+   * Returns the lowercase, trimmed display text of a menu option.
+   * Uses the rendered text so matching mirrors what the user sees and what screen readers announce,
+   * matching native HTML <select> behavior.
+   * @private
+   * @param {HTMLElement} option - The menu option element.
+   * @returns {string}
+   */
+  _getOptionDisplayText(option) {
+    return (option.innerText || option.textContent || '').trim().toLowerCase();
+  }
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/src/auro-select.js:929]: Addressed.

File: components/select/src/auro-select.js [hunk 4]
```diff
@@ -880,41 +902,88 @@ export class AuroSelect extends AuroElement {
...
 
-      // Select the new option in the menu
-      const newOption = letterOptions[index];
-      this.menu.updateActiveOption(newOption);
+    if (match) {
+      if (!this.dropdown.isPopoverVisible) {
+        this.dropdown.show();
+      }
+      this.menu.updateActiveOption(match);
     }
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/src/auro-select.js:1004]: Keeping the current behavior intentionally — it mirrors native HTML `<select>` and the WAI-ARIA Listbox pattern, both of which preserve the buffer on no-match so intermediate keystrokes during multi-character prefix matching don't reset progress. Example: typing `s`-`a`-`n`-`t`-`a` for "Santa Barbara"; the intermediate `t` produces no match against any `sant*` option, but clearing here would force the user to retype `sant` after the timeout.

The Space-as-toggle trap you're describing is real but bounded — Space recovers after `typeaheadTimeoutMs` (default 500ms) and `Escape` clears the buffer immediately (wired in a separate commit). Documented the buffer-clear-on-Escape behavior in the Customize → Type-Ahead section so users learn the escape hatch.

File: components/select/src/auro-select.js [hunk 5]
```diff
@@ -880,41 +902,88 @@ export class AuroSelect extends AuroElement {
...
-    // Get a lowercase version of the key pressed
-    const key = _key.toLowerCase();
-
-    // Calculate how many times the same letter has been pressed
-    this.sameLetterTimes = key === this.lastLetter ? this.sameLetterTimes + 1 : 0;
+    // No enabled options to match against — clear any stale buffer/timer so
+    // Space stays a bib toggle and the next keystroke after fresh options load
+    // starts cleanly. Checked BEFORE mutating the buffer.
+    const options = (this.menu && this.menu.options ? this.menu.options : []).filter((option) => !option.disabled);
+    if (!options.length) {
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/src/auro-select.js]: Addressed.

File: components/select/docs/partials/keyboard-behavior/keyEvents.md [hunk 6]
```diff
@@ -196,5 +196,23 @@
         </div>
       </td>
     </tr>
+    <tr>
+      <td>Space</td>
+      <td>-</td>
+      <td>Collapsed or Expanded</td>
+      <td>Trigger element</td>
+      <td>
+        Overrides the inherited Space behavior while a type-ahead buffer is active: instead of toggling the bib, Space extends the buffer — the <code>focused</code> option advances if the buffer matches an enabled option, and the bib state is unchanged. See the <a href="customize.html#typeAhead">Type-Ahead</a> section on the Customize page.
```
[the remark this answers is not in the document]
Comment (jordanjones243) [reply components/select/docs/partials/keyboard-behavior/keyEvents.md]: Addressed.

**Review by jason-capsule42 — DISMISSED** [review]
Tests are failing and docs/post-mortem/<ticketNumber> file is missing.

File: components/select/src/auro-select.js [hunk 7]
```diff
@@ -896,41 +918,90 @@ export class AuroSelect extends AuroElement {
...
-    // Set last letter for tracking
-    this.lastLetter = key;
+    const key = _key.toLowerCase();
 
-    // Get all the options that start with the last letter pressed
-    const letterOptions = this.menu.options.filter((option) => {
-      const optionText = option.value || '';
-      return optionText.toLowerCase().startsWith(this.lastLetter);
-    });
+    // Reset the buffer after a period of inactivity
```
Comment (chrisfalaska) [inline components/select/src/auro-select.js:976]: Suggestion: Route the timeout-fired buffer reset through `_clearTypeaheadBuffer`

Single source of truth -- every other place that clears the buffer already goes through `_clearTypeaheadBuffer`.

Currently the timer callback reassigns directly, which is fine today but invites drift if `_clearTypeaheadBuffer` ever grows additional responsibilities (e.g. dispatching a future "type-ahead ended" event).

```
this._typeaheadTimeout = setTimeout(() => {
  this._clearTypeaheadBuffer();
}, this.typeaheadTimeoutMs);
```
(_clearTypeaheadBuffer already nulls _typeaheadTimeout after clearing it, which is correct here — the timer has fired, so clearTimeout on a fired timer is a harmless no-op.)

File: components/select/test/auro-select.test.js [hunk 8]
```diff
@@ -3518,6 +3497,435 @@ function runTest(mobileView) {
...
+          await elementUpdated(el);
+          expect(el._typeaheadTimeout).to.not.equal(null);
+
+          el._clearTypeaheadBuffer();
+          expect(el._typeaheadTimeout).to.equal(null);
+          expect(el.typeaheadBuffer).to.equal('');
+        });
       });
     });
   });
```
Comment (chrisfalaska) [inline components/select/test/auro-select.test.js:3943]: Add multiselect + type-ahead smoke test

Type-ahead calls `menu.updateActiveOption` (active-descendant only, not selection) so it should work in multiselect, but it isn't asserted. Below the existing "no enabled options, Space stays toggle" test:

```
  it('cycles active option via type-ahead in multiselect mode without altering selection', async () => {
    const el = await multiSelectFixture(); // existing helper used elsewhere in this file
    await elementUpdated(el);

    el.focus();
    el.dropdown.show();
    await elementUpdated(el);

    const previouslySelected = [...(el.optionSelected || [])];

    // Press 'b' — should move active option to first option starting with 'b'
    el.dispatchEvent(new KeyboardEvent('keydown', { key: 'b', bubbles: true, composed: true }));
    await elementUpdated(el);

    expect(el.menu.optionActive).to.exist;
    expect(el.menu.optionActive.textContent.trim().toLowerCase().startsWith('b')).to.be.true;

    // Selection state is unchanged — type-ahead must not toggle checks in multiselect
    expect(el.optionSelected).to.deep.equal(previouslySelected);
  });
```

File: components/select/test/auro-select.test.js
Comment (chrisfalaska) [inline components/select/test/auro-select.test.js]: Clean up the pre-existing synthetic auroMenu-selectedOption test

Pre-dates this PR but is now stale — v6 dropped `event.detail.stringValue/detail.options`.

The test passes today only because the production handler reads from  `this.menu.*`, so the synthetic detail is effectively ignored. Either remove or rewrite to drive a real selection. Out of scope for this PR — file as a follow-up issue.

Suggested follow-up replacement:

```
  it('optionSelected is undefined when menu has no items', async () => {
    const el = await emptyMenuFixture();
    await elementUpdated(el);
    expect(el.optionSelected).to.equal(undefined);
  });
```

**Review by chrisfalaska — CHANGES_REQUESTED** [review]

**Comment by chrisfalaska** [comment]
Optional: extend `selectUtils.js` instead of inlining the filter

**File:** `components/select/src/selectUtils.js`

**Why:** Two callers (v6 strategy + new type-ahead path) now do
`(menu?.options || []).filter(...)` with different predicates.
Co-locating both keeps the "safely get options" pattern in one place.

Non-blocking — the inline filter with its current comment is also
defensible.

```js
/**
 * Returns the "active" (selectable + visible) options for type-ahead navigation.
 * Excludes disabled, hidden, and static options (e.g. `static nomatch`
 * placeholders) so the type-ahead cursor doesn't land on a non-actionable item.
 *
 * @param {HTMLElement | null | undefined} menu
 * @returns {Array<HTMLElement>}
 */
export function getActiveOptions(menu) {
  return (menu?.options || []).filter((option) => option.isActive);
}
```

Then in `auro-select.js`:

```js
const options = getActiveOptions(this.menu);
```

(with `import { getActiveOptions } from './selectUtils.js';` added.)

File: components/select/src/auro-select.js (hunk 7 above)
Comment (jordanjones243) [reply components/select/src/auro-select.js:976]: Addressed.

File: components/select/test/auro-select.test.js (hunk 8 above)
Comment (jordanjones243) [reply components/select/test/auro-select.test.js:3943]: Addressed.

File: components/select/src/selectUtils.js [hunk 9]
```diff
@@ -12,3 +12,18 @@
...
+ *
+ * Uses auro-menuoption's `isActive` getter, which excludes disabled, hidden,
+ * and static options (e.g. `static nomatch` placeholders) so the type-ahead
+ * cursor never lands on a non-actionable item. Same empty-safe handling as
+ * `getEnabledOptions`.
+ *
+ * @param {HTMLElement | null | undefined} menu - The auro-menu element.
+ * @returns {Array<HTMLElement>} Active options, empty array when none.
+ */
+export function getActiveOptions(menu) {
```
Comment (chrisfalaska) [inline components/select/src/selectUtils.js:27]: One missing thing: `selectUtils.test.js` was not updated. 

`getEnabledOptions` has three unit tests (happy path, empty menu, null/undefined menu). getActiveOptions has zero -- it's only exercised indirectly via the type-ahead integration tests.

 Suggested add (mirrors the existing structure):

```
  describe('getActiveOptions', () => {
    it('returns only active options (excludes disabled, hidden, static)', async () => {
      const menu = await fixture(html`
        <auro-menu>
          <auro-menuoption value="a">Apples</auro-menuoption>
          <auro-menuoption value="b" disabled>Bananas</auro-menuoption>
          <auro-menuoption value="c" hidden>Cherries</auro-menuoption>
          <auro-menuoption value="d" static>No match</auro-menuoption>
          <auro-menuoption value="e">Elderberries</auro-menuoption>
        </auro-menu>
      `);

      const active = getActiveOptions(menu);
      expect(active.map((opt) => opt.value)).to.deep.equal(['a', 'e']);
    });

    it('returns an empty array when the menu has no items', async () => {
      const menu = await fixture(html`<auro-menu></auro-menu>`);
      expect(getActiveOptions(menu)).to.deep.equal([]);
    });

    it('returns an empty array when the menu is null or undefined', () => {
      expect(getActiveOptions(undefined)).to.deep.equal([]);
      expect(getActiveOptions(null)).to.deep.equal([]);
    });
  });
```

- Also be sure to add `getActiveOptions` to the import at the top of the test file.

**Review by chrisfalaska — CHANGES_REQUESTED** [review]

**Review by chrisfalaska — APPROVED** [review]
Looks great! 🚀
